proposals/0000-compiler-settings.md

```diff
@@ -0,0 +1,284 @@
+involves instead:
+
+1. Enabling warningsAsErrors for the new feature on the specific file.
+
+2. Fixing all of the errors in the file.
+
+3. Commiting the updated file into main.
+
+Since one is only updating one file at a time and using warningsAsErrors, the
+updates can be done incrementally on a per file basis, backsliding cannot occur
+in the file, and most importantly main is never in an intermediate, partially
+updated state.
+
+## Proposed solution
+
+We propose introducing a new top level statement called `compilerSettings` that
+takes a list of enums of type `CompilerSetting`. Each enum case's associated
+values represent arguments that would be passed as an argument to the compiler
+flag on the command line. The compiler upon parsing a `compilerSettings`
+statement updates its internal state to set the appropriate settings before
+compiling the rest of the file. An example of such a `compilerSettings`
+statement is the following:
+
+```swift
+compilerSettings [
+    .warningAsError(.concurrency),
+    .strictConcurrency(.complete),
+    .enableUpcomingFeature(.existentialAny),
+    .defaultIsolation(MainActor.self),
+]
+```
+
+By specifying compiler options in such a manner, we are able to solve all of the
+problems above:
+
+1. The programmer can avoid module splitting when they wish to use compiler
+   flags only on a specific file.
+
+2. A programmer can specify options on a per file basis instead of only on a per
+   module basis allowing for warningsAsError to be used per file instead of per
+   module migration to use new features.
+
+As an additional benefit since enums are being used, code completion can make
+the compiler options discoverable to the user in the IDE.
+
+## Detailed design
+
+We will update the swift grammer to allow for a new top level statement called
+`compilerSettings` that takes a list of expressions:
+
+```text
+top-level-statement ::= 'compilerSettings' '[' (expr ',')* expr ','? ']'
+```
+
+Each expression passed to `compilerSettings` is an instance of the enum
+`CompilerSetting` that specifies a command line option that is to be applied to
+the file. `CompilerSetting` will be a resilient enum defined in the standard
+library that provides the following API:
+
+```
+public enum CompilerSetting {
+case enableUpcomingFeature(UpcomingFeatureCompilerSetting)
+case enableExperimentalFeature(ExperimentalFeatureCompilerSetting)
+case warningAsError(WarningAsErrorCompilerSetting)
+case defaultIsolation(Actor.Type)
+case strictConcurrency(StrictConcurrencyCompilerSetting)
+}
+```
+
+We purposely make `CompilerSetting` resilient so additional kinds of compiler
+settings can be added over time without breaking ABI.
+
+`CompilerSetting` provides an enum case for each command line option and each
+case is able to provide associated values to specify arguments that would
+normally be passed on the command line to the option. These associated values
+can be one of:
+
+* an already existing type (e.x.: `Actor.Type`)
```

Comment:
I think `e.g.` is a more common abbreviation than `e.x.`, but IMO in technical writing it's better to spell out "for example"